---
title: "Filtering Links by DARs and Gex"
author: 
- name: "Maria Lucia Romero Rivero"
  email: MLROMEROR@clinic.cat
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
    fig_captions: yes

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Pre-processing 

## Load package


```{r message=FALSE, warning=FALSE}
library(RColorBrewer)
library(Signac)
library(Seurat)
library(GenomicRanges)
library(future)
library(harmony)
library(EnsDb.Hsapiens.v86)
library(stringr)
library(dplyr)
library(ggplot2)
library(patchwork)
library(kableExtra)
library(tidyverse)
library(BSgenome.Hsapiens.UCSC.hg38)
library(plyr)
library(biomaRt)
library(BiocManager)
library(GenomicRanges)
set.seed(123)
```


## Parameters

```{r}
# Paths
path_to_obj <- ("~/Documents/multiome_tonsil_Lucia/results/R_objects/19.tonsil_multiome_bcell_level_1_Genwide_Linkpeak.rds")


path_to_obj_save_dars_dea <- ("~/Documents/multiome_tonsil_Lucia/results/R_objects/19.tonsil_multiome_bcell_level_1_Genomewide_Linkpeak_DARs_DEA.rds")

path_to_save_DARs<-("~/Documents/multiome_tonsil_Lucia/results/tables/19.df_DARs_level_1.csv")
path_to_save_join_link_DARs <- ("~/Documents/multiome_tonsil_Lucia/results/tables/19.df_joint_link_DARs_level_1_Genwide.csv")
path_to_save_df_join_link_DARs_DEA<-  ("~/Documents/multiome_tonsil_Lucia/results/tables/19.df_joint_link_DARs_DEA_level_1_Genwide.csv")

path_to_save_markers<-("~/Documents/multiome_tonsil_Lucia/results/tables/18_tonsil_markers_level_1_namecluster.csv")

```


## Load data

```{r}
tonsil_wnn_bcell <- readRDS(path_to_obj)
```

# RenameIdents

```{r}
Idents(tonsil_wnn_bcell)<-"wsnn_res.0.05"
new.cluster.ids <- c("NBC","MBC", "GC", "GC","PC")

names(new.cluster.ids) <- levels(tonsil_wnn_bcell)
tonsil_wnn_bcell <- RenameIdents(tonsil_wnn_bcell, new.cluster.ids)
tonsil_wnn_bcell[["annotation_level_1"]] <- Idents(object = tonsil_wnn_bcell)


DimPlot(tonsil_wnn_bcell,label = TRUE, reduction = "wnn.umap", pt.size = 0.5,cols = c("#a6cee3", "#1f78b4","#b2df8a", 
             "#33a02c", "#fb9a99","#e31a1c")) + ggtitle("Joint UMAP")+NoLegend()
```


# Join DARs

```{r}
df_tonsil_DARs<-FindAllMarkers(object = tonsil_wnn_bcell, only.pos = TRUE,assay = "ATAC")

```

```{r}
dars_nbc<-df_tonsil_DARs[df_tonsil_DARs$cluster=="NBC",]

dars_gcbc<-df_tonsil_DARs[df_tonsil_DARs$cluster=="GC",]
dars_mbc<-df_tonsil_DARs[df_tonsil_DARs$cluster=="MBC",]
dars_pc<-df_tonsil_DARs[df_tonsil_DARs$cluster=="PC",]
df_dar_celltype<-data.frame(cell_type=c("NBC","GCBC","GCBC","PC"), num_dars=c(nrow(dars_nbc),nrow(dars_gcbc),nrow(dars_mbc),nrow(dars_pc)))
```


```{r}
coord_link_DARs<-str_split_fixed(df_tonsil_DARs$gene, "-", 3)
colnames(coord_link_DARs)<-c("chrom","start","end")
df_tonsil_DARs<-cbind(df_tonsil_DARs,coord_link_DARs)
write_csv(df_tonsil_DARs,path_to_save_DARs)

```

```{r}
df_tonsil_DARs<-read.csv(path_to_save_DARs)

DT::datatable(head(df_tonsil_DARs))
```


**From dataframe to Granges**

```{r}
names(df_tonsil_DARs)[c(8,7)]<-c("seqnames","peak")

 #Joint links to DARs

df_linkpeak<-as.data.frame(Links(tonsil_wnn_bcell))
df_links_DARs<-inner_join(df_linkpeak,df_tonsil_DARs,by="peak")
names(df_links_DARs)[c(1,2,3,18,19)]<-c("seqnames","start","end","start.peak","end.peak")
df_links_DARs$seqnames.y<-NULL
write_csv(df_links_DARs,path_to_save_join_link_DARs)

```


## Create another Seurat Object with Links in DARs


```{r}
tonsil_wnn_bcell_links_DARs<-tonsil_wnn_bcell
Links(tonsil_wnn_bcell_links_DARs)<-makeGRangesFromDataFrame(df_links_DARs,
                                                   keep.extra.columns=FALSE,
                                                   ignore.strand=FALSE,
                                                   seqinfo=NULL,
                                                   seqnames.field=c("seqnames", "seqname","chromosome", 
                                                                    "chrom","chr","chromosome_name","seqid"),
                                                   start.field="start",
                                                   end.field=c("end", "stop"),
                                                   strand.field="strand",
                                                   starts.in.df.are.0based=FALSE)

values(Links(tonsil_wnn_bcell_links_DARs))<-DataFrame(score=df_links_DARs$score,gene=df_links_DARs$gene, cluster=df_links_DARs$cluster,peak=df_links_DARs$peak,zscore=df_links_DARs$zscore,pvalue=df_links_DARs$pvalue,p_val.dars=df_links_DARs$p_val,pct.1=df_links_DARs$pct.1,pct.2=df_links_DARs$pct.2, avg_log2FC=df_links_DARs$avg_log2FC,start.peak=df_links_DARs$start.peak,end.peak=df_links_DARs$end.peak)

```



## Coverage plot

*`coverage_extend{r}`is a function that create a coverageplot for a certain gene and region using different upstream and downstream extension region. *

```{r}
coverage_extend <- function(x,y,seuratobject){purrr::map(y, function(y) {
  
  p <- CoveragePlot(
  object = seuratobject,
  region = x,
  features = x,
  expression.assay = "RNA",
  idents = idents.plot,
  extend.upstream = y,
  extend.downstream = y
  #tile = TRUE
  )
  p & scale_fill_manual(values = cols_cluster)
})}
```

```{r}
ranges.show <- StringToGRanges("chr3-187721377-187745725")
cols_cluster <- c("#a6cee3", "#1f78b4","#b2df8a",  
             "#33a02c","#e31a1c")

idents.plot <- c("MBC", "NBC","GC", "PC")
coverage_extend("BCL6",c(0,2000,10000,1e+7),tonsil_wnn_bcell)
```

```{r}
coverage_extend("BCL6",c(0,2000,10000,1e+7),tonsil_wnn_bcell_links_DARs)
```

```{r}
coverage_extend("PRDM1",c(0,2000,10000,100000,1e+6),tonsil_wnn_bcell_links_DARs)

```




# Joint DEA


```{r}
tonsil_markers<-FindAllMarkers(object = tonsil_wnn_bcell, only.pos = TRUE,assay = "RNA")
#write_csv(tonsil_markers,path_to_save_markers)
```


```{r}
df_links_DARs_DEA<-inner_join(df_links_DARs,tonsil_markers,by=c("gene","cluster"))
names(df_links_DARs_DEA)[c(11,12,13,14,15,19,20,21,22,23)]<-c("p_val.dars","avg_log2FC.dars","pct.1.dars" ,"pct.2.dars" ,"p_val_adj.dars","p_val.rna", "avg_log2FC.rna","pct.1.rna" ,"pct.2.rna", "p_val_adj.rna")
head(df_links_DARs_DEA)
#write_csv(df_links_DARs_DEA,path_to_save_df_join_link_DARs_DEA)

```

```{r}
kbl(head(df_linkpeak),caption = "Table of all Link peaks of tonsil leve 1 object") %>%
  kable_paper("striped", full_width = F)
```


```{r}
kbl(tonsil_markers,caption = "Table of the markers") %>%
  kable_paper("striped", full_width = F)
```


```{r}
kbl(df_tonsil_DARs,caption = "Table of the DARs") %>%
  kable_paper("striped", full_width = F)
```
```{r}
kbl(df_links_DARs,caption = "Table of the join of the all links peaks and the DARs") %>%
  kable_paper("striped", full_width = F)
```
```{r}
kbl(df_links_DARs_DEA,caption = "Table of the filtered links peaks by DARs and GEx") %>%
  kable_paper("striped", full_width = F)
```

```{r}
tonsil_wnn_bcell_links_DARs_DEA<-tonsil_wnn_bcell
Links(tonsil_wnn_bcell_links_DARs_DEA)<-makeGRangesFromDataFrame(df_links_DARs_DEA,
                                                   keep.extra.columns=FALSE,
                                                   ignore.strand=FALSE,
                                                   seqinfo=NULL,
                                                   seqnames.field=c("seqnames", "seqname","chromosome", 
                                                                    "chrom","chr","chromosome_name","seqid"),
                                                   start.field="start",
                                                   end.field=c("end", "stop"),
                                                   strand.field="strand",
                                                   starts.in.df.are.0based=FALSE)

values(Links(tonsil_wnn_bcell_links_DARs_DEA))<-DataFrame(score=df_links_DARs_DEA$score,
                                            gene=df_links_DARs_DEA$gene,
                                            cluster=df_links_DARs_DEA$cluster,
                                            peak=df_links_DARs_DEA$peak,
                                            zscore=df_links_DARs_DEA$zscore,
                                            pvalue=df_links_DARs_DEA$pvalue,
                                            p_val.dars=df_links_DARs_DEA$p_val.dars,
                                            pct.1.dars=df_links_DARs_DEA$pct.1.dars,
                                            pct.2.dars=df_links_DARs_DEA$pct.2.dars, 
                                            avg_log2FC.dars=df_links_DARs_DEA$avg_log2FC.dars,
                                            start.peak=df_links_DARs_DEA$start.peak,
                                            end.peak=df_links_DARs_DEA$end.peak,
                                            p_val.rna=df_links_DARs_DEA$p_val.rna, 
                                            avg_log2FC.rna=df_links_DARs_DEA$avg_log2FC.rna, 
                                            pct.1.rna=df_links_DARs_DEA$pct.1.rna, 
                                            pct.2.rna=df_links_DARs_DEA$pct.2.rna, 
                                            p_val_adj.rna=df_links_DARs_DEA$p_val_adj.rna)
table(df_links_DARs_DEA$gene=="BCL6",df_links_DARs_DEA$cluster)

```
```{r}
table(df_links_DARs_DEA$gene=="PRDM1",df_links_DARs_DEA$cluster)

```

```{r}
ranges.show <- StringToGRanges("chr3-187721377-187745725")
cols_cluster <- c("#a6cee3", "#1f78b4","#b2df8a",  
             "#33a02c","#e31a1c")

idents.plot <- c("MBC", "NBC","GC", "PC")
coverage_extend("BCL6",c(0,2000,10000,1e+7),tonsil_wnn_bcell_links_DARs_DEA)
```

# Save

```{r eval=FALSE, include=FALSE}
saveRDS(tonsil_wnn_bcell, path_to_obj_save)
saveRDS(tonsil_wnn_bcell_links_DARs_DEA, path_to_obj_save_dars_dea)

```

# Session info

```{r}
sessionInfo()
```